src/darwin/Framework/CHIP/MTRDiagnosticLogsDownloader.mm

```diff
@@ -0,0 +1,630 @@
+}
+@end
+
+@implementation MTRDiagnosticLogsDownloader
+- (instancetype)init
+{
+    assertChipStackLockedByCurrentThread();
+
+    if (self = [super init]) {
+        _downloads = [[Downloads alloc] init];
+        _bridge = new DiagnosticLogsDownloaderBridge(self);
+        if (_bridge == nullptr) {
+            MTR_LOG_ERROR("Error: %@", kErrorInitDiagnosticLogsDownloader);
+            return nil;
+        }
+    }
+    return self;
+}
+
+- (void)dealloc
+{
+    if (_bridge) {
+        delete _bridge;
+        _bridge = nil;
+    }
+    _downloads = nil;
+}
+
+- (chip::bdx::BDXTransferServerDelegate *)getBridge
+{
+    return _bridge;
+}
+
+- (void)downloadLogFromNodeWithID:(NSNumber *)nodeID
+                       controller:(MTRDeviceController *)controller
+                             type:(MTRDiagnosticLogType)type
+                          timeout:(NSTimeInterval)timeout
+                            queue:(dispatch_queue_t)queue
+                       completion:(void (^)(NSURL * _Nullable url, NSError * _Nullable error))completion;
+{
+    assertChipStackLockedByCurrentThread();
+
+    uint16_t timeoutInSeconds = 0;
+    if (timeout <= 0) {
+        timeoutInSeconds = 0;
+    } else if (timeout > UINT16_MAX) {
+        MTR_LOG_INFO("Warning: timeout is too large. It will be truncated to UINT16_MAX.");
+        timeoutInSeconds = UINT16_MAX;
+    } else {
+        timeoutInSeconds = static_cast<uint16_t>(timeout);
+    }
+
+    // This block is always called when a download is finished.
+    auto done = ^(Download * finishedDownload) {
+        [controller asyncDispatchToMatterQueue:^() {
+            [self->_downloads remove:finishedDownload];
+
+            if (timeoutInSeconds > 0) {
+                self->_bridge->CancelBDXTransferTimeout(finishedDownload);
+            }
+        } errorHandler:nil];
+    };
+
+    auto fabricIndex = @(controller.fabricIndex);
+    auto download = [_downloads add:type fabricIndex:fabricIndex nodeID:nodeID queue:queue completion:completion done:done];
+    VerifyOrReturn(nil != download,
+        dispatch_async(queue, ^{ completion(nil, [MTRError errorForCHIPErrorCode:CHIP_ERROR_INTERNAL]); }));
+
+    auto interactionModelDone = ^(MTRDiagnosticLogsClusterRetrieveLogsResponseParams * _Nullable response, NSError * _Nullable error) {
+        [download checkInteractionModelResponse:response error:error];
+    };
+
+    auto * device = [controller deviceForNodeID:nodeID];
+    auto * cluster = [[MTRClusterDiagnosticLogs alloc] initWithDevice:device endpointID:@(kDiagnosticLogsEndPoint) queue:queue];
+
+    auto * params = [[MTRDiagnosticLogsClusterRetrieveLogsRequestParams alloc] init];
+    params.intent = @(type);
+    params.requestedProtocol = @(MTRDiagnosticLogsTransferProtocolBDX);
+    params.transferFileDesignator = download.fileDesignator;
+
+    [cluster retrieveLogsRequestWithParams:params expectedValues:nil expectedValueInterval:nil completion:interactionModelDone];
```

Comment:
Using public framework APIs from inside the Matter queue (per assert above) is _really_ dangerous, since some of those do dispatch_sync things.  Have you audited this codepath to make sure it doesn't?

Again, this is why we _must_ have a test in CI exercising this.  This is extremely fragile.

Reply:
I have not seen any deadlock. Would you prefer me to dispatch that onto the user provided queue ?